<a href="https://colab.research.google.com/github/shintafiaa/Kampus-Merdeka-ML-Terapan/blob/main/7_Content_Based_Filtering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Recommendation System on Delivery Food App Case using Content Based Filtering**

Pada latihan ini sistem rekomendasi menggunakan cara content-based filtering. Kita akan membangun sistem rekomendasi sederhana berdasarkan jenis masakan yang disediakan restoran. Data yang dibutuhkan adalah nama restoran dan cuisine. Kita akan menghitung kesamaan (similarity) cuisine dan nama restoran kemudian membuat rekomendasi berdasarkan kesamaan ini.

download dataset 

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip

--2021-11-14 10:29:13--  https://archive.ics.uci.edu/ml/machine-learning-databases/00232/RCdata.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 39674 (39K) [application/x-httpd-php]
Saving to: ‘RCdata.zip’

RCdata.zip          100%[===================>]  38.74K  --.-KB/s    in 0.1s    

2021-11-14 10:29:13 (298 KB/s) - ‘RCdata.zip’ saved [39674/39674]



File RCdata.zip tersimpan di storage google colab. Lalu, unzip file tersebut.

In [3]:
!unzip /content/RCdata.zip

Archive:  /content/RCdata.zip
  inflating: chefmozaccepts.csv      
  inflating: chefmozcuisine.csv      
  inflating: chefmozhours4.csv       
  inflating: chefmozparking.csv      
  inflating: geoplaces2.csv          
  inflating: rating_final.csv        
  inflating: README                  
  inflating: usercuisine.csv         
  inflating: userpayment.csv         
  inflating: userprofile.csv         


9 berkas terdiri dari:

**Restaurants:**
- chefmozaccepts.csv
- chefmozcuisine.csv
- chefmozhours4.csv
- chefmozparking.csv
- geoplaces2.csv

**Consumers:**
- usercuisine.csv
- userpayment.csv
- userprofile.csv

**User-Item-Rating:**
- rating_final.csv

Baca masing-masing berkas

In [4]:
import pandas as pd
 
accepts = pd.read_csv('/content/chefmozaccepts.csv')
cuisine = pd.read_csv('/content/chefmozcuisine.csv')
hours = pd.read_csv('/content/chefmozhours4.csv')
parking = pd.read_csv('/content/chefmozparking.csv')
geo = pd.read_csv('/content/geoplaces2.csv', encoding = "ISO-8859-1")
usercuisine = pd.read_csv('/content/usercuisine.csv')
payment = pd.read_csv('/content/userpayment.csv')
profile = pd.read_csv('/content/userprofile.csv')
rating = pd.read_csv('/content/rating_final.csv')
 
print('Jumlah data pembayaran yang diterima restoran: ', len(accepts.placeID.unique()))
print('Jumlah data masakan pada restoran: ', len(cuisine.placeID.unique()))
print('Jumlah data waktu buka restoran: ', len(hours.placeID.unique()))
print('Jumlah data lokasi restoran: ', len(geo.placeID.unique()))
print('Jumlah data masakan pengguna: ', len(usercuisine.userID.unique()))
print('Jumlah data profil pengguna: ', len(profile.userID.unique()))
print('Jumlah data penilaian yang diberikan pengguna: ', len(rating.userID.unique()))
print('Jumlah data penilaian restoran: ', len(rating.placeID.unique()))

Jumlah data pembayaran yang diterima restoran:  615
Jumlah data masakan pada restoran:  769
Jumlah data waktu buka restoran:  694
Jumlah data lokasi restoran:  130
Jumlah data masakan pengguna:  138
Jumlah data profil pengguna:  138
Jumlah data penilaian yang diberikan pengguna:  138
Jumlah data penilaian restoran:  130


#Univariate Exploratory Data Analysis

Variabel-variabel pada dataset:
- accepts : merupakan jenis pembayaran yang diterima pada restoran tertentu.
- cuisine : merupakan jenis masakan yang disajikan pada restoran.
- hours : merupakan jadwal buka dan tutupnya restoran.
- parking : merupakan ketersediaan tempat parkir pada restoran.
- geo : merupakan letak restoran.
- usercuisine : merupakan jenis masakan dari data pengguna.
- payment : merupakan jenis pembayaran yang dipakai pengguna.
- profile : merupakan data profil pengguna.

**Eksplorasi Data Accepts**

In [5]:
accepts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1314 entries, 0 to 1313
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   1314 non-null   int64 
 1   Rpayment  1314 non-null   object
dtypes: int64(1), object(1)
memory usage: 20.7+ KB


melihat ada berapa banyak entri yang unik berdasarkan placeID dan jenis-jenis pembayaran unik yang diterima

In [6]:
print('Banyak data: ', len(accepts.placeID.unique()))
print('Jenis Pembayaran yang diterima: ', accepts.Rpayment.unique())

Banyak data:  615
Jenis Pembayaran yang diterima:  ['cash' 'VISA' 'MasterCard-Eurocard' 'American_Express' 'bank_debit_cards'
 'checks' 'Discover' 'Carte_Blanche' 'Diners_Club' 'Visa'
 'Japan_Credit_Bureau' 'gift_certificates']


Data accept menunjukkan terdapat 1314 entri,namun hanya terdapat 615 data setelah disaring dengang fungsil unique()

**Eksplorasi Data Cuisine**

Variabel cuisine menunjukkan jenis masakan yang tersedia di restoran.

In [7]:
cuisine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   916 non-null    int64 
 1   Rcuisine  916 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.4+ KB


In [8]:
print('Banyak tipe masakan: ', len(cuisine.Rcuisine.unique()))
print('Tipe masakan: ', cuisine.Rcuisine.unique())

Banyak tipe masakan:  59
Tipe masakan:  ['Spanish' 'Italian' 'Latin_American' 'Mexican' 'Fast_Food' 'Burgers'
 'Dessert-Ice_Cream' 'Hot_Dogs' 'Steaks' 'Asian' 'International'
 'Mongolian' 'Vegetarian' 'Brazilian' 'Cafe-Coffee_Shop' 'Cafeteria'
 'Contemporary' 'Deli-Sandwiches' 'Diner' 'Japanese' 'Sushi' 'Seafood'
 'Chinese' 'Bar' 'Bar_Pub_Brewery' 'Pizzeria' 'Mediterranean' 'American'
 'Family' 'Caribbean' 'African' 'Breakfast-Brunch' 'Regional' 'Afghan'
 'Bakery' 'Game' 'Armenian' 'Vietnamese' 'Korean' 'Thai' 'Barbecue'
 'Polish' 'Dutch-Belgian' 'French' 'German' 'Southwestern' 'Persian'
 'Ethiopian' 'Juice' 'Soup' 'Continental-European' 'Greek' 'Southern'
 'Eastern_European' 'California' 'Bagels' 'Turkish' 'Organic-Healthy'
 'Fine_Dining']


Data cuisine digunakan untuk memprediksi top-N rekomendasi bagi pengguna

**Eksplorasi Data Profile**

In [9]:
print(profile.shape)

(138, 19)


Terdapat 138 baris dan 19 kolom pada data profile.

In [10]:
profile.head()

userID   latitude   longitude smoker  ...  color weight  budget height
0  U1001  22.139997 -100.978803  false  ...  black     69  medium   1.77
1  U1002  22.150087 -100.983325  false  ...    red     40     low   1.87
2  U1003  22.119847 -100.946527  false  ...   blue     60     low   1.69
3  U1004  18.867000  -99.183000  false  ...  green     44  medium   1.53
4  U1005  22.183477 -100.959891  false  ...  black     65  medium   1.69

[5 rows x 19 columns]

**Eksplorasi Data Rating**

In [11]:
rating.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

Data rating terdiri dari 5 kolom dengan tiga kategori rating.

Penjelasan masing-masing variabel:
- userID, merupakan identitas pengguna.
- placeID, merupakan identitas restoran.
- Rating, merupakan data rating untuk restoran.
- Food_rating, merupakan data rating untuk makanan atau masakan di restoran tersebut.
- Service_rating, merupakan data layanan restoran tersebut.

Melihat distribusi data pada data rating.

In [12]:
rating.describe()

placeID       rating  food_rating  service_rating
count    1161.000000  1161.000000  1161.000000     1161.000000
mean   134192.041344     1.199828     1.215332        1.090439
std      1100.916275     0.773282     0.792294        0.790844
min    132560.000000     0.000000     0.000000        0.000000
25%    132856.000000     1.000000     1.000000        0.000000
50%    135030.000000     1.000000     1.000000        1.000000
75%    135059.000000     2.000000     2.000000        2.000000
max    135109.000000     2.000000     2.000000        2.000000

Nilai maksimum rating adalah 2 dan nilai minimumnya adalah 0. 

Artinya, skala rating berkisar antara 0 hingga 2.

Melihat berapa pengguna yang memberikan rating, jumlah restoran, dan jumlah rating.

In [13]:
print('Jumlah userID: ', len(rating.userID.unique()))
print('Jumlah placeID: ', len(rating.placeID.unique()))
print('Jumlah data rating: ', len(rating))

Jumlah userID:  138
Jumlah placeID:  130
Jumlah data rating:  1161


#Data Preprocessing

**Menggabungkan 5 Data Restoran**

Identifikasi jumlah seluruh restoran pada dataset. Menggunakan placeID yang unik sebagai acuan dalam penggabungan.

In [14]:
import numpy as np
 
# Menggabungkan seluruh placeID pada kategori Restaurant
resto_all = np.concatenate((
    accepts.placeID.unique(),
    cuisine.placeID.unique(),
    hours.placeID.unique(),
    parking.placeID.unique(),
    geo.placeID.unique()
))
 
# Mengurutkan data dan menghapus data yang sama
resto_all = np.sort(np.unique(resto_all))
 
print('Jumlah seluruh data restoran berdasarkan placeID: ', len(resto_all))

Jumlah seluruh data restoran berdasarkan placeID:  938


**Menggabungkan 3 Data User**

In [15]:
# Menggabungkan seluruh userID
user_all = np.concatenate((
    usercuisine.userID.unique(),
    payment.userID.unique(),
    profile.userID.unique()
))
 
# Menghapus data yang sama kemudian mengurutkannya
user_all = np.sort(np.unique(user_all)) 
 
print('Jumlah seluruh user: ', len(user_all))

Jumlah seluruh user:  138


**Mengetahui Jumlah Rating**

In [16]:
# Menggabungkan file accepts, geo, parking, hours ke dalam dataframe resto_info 
resto_info = pd.concat([accepts, geo, parking, hours])
 
# Menggabungkan dataframe rating dengan resto_info berdasarkan nilai placeID
resto = pd.merge(rating, resto_info , on='placeID', how='left')
resto

userID  placeID  rating  ...  parking_lot         hours                  days
0     U1077   135085       2  ...          NaN           NaN                   NaN
1     U1077   135085       2  ...          NaN           NaN                   NaN
2     U1077   135085       2  ...       public           NaN                   NaN
3     U1077   135085       2  ...          NaN  00:00-00:00;  Mon;Tue;Wed;Thu;Fri;
4     U1077   135085       2  ...          NaN  00:00-00:00;                  Sat;
...     ...      ...     ...  ...          ...           ...                   ...
8073  U1068   132660       0  ...          NaN           NaN                   NaN
8074  U1068   132660       0  ...         none           NaN                   NaN
8075  U1068   132660       0  ...          NaN  00:00-23:30;  Mon;Tue;Wed;Thu;Fri;
8076  U1068   132660       0  ...          NaN  00:00-23:30;                  Sat;
8077  U1068   132660       0  ...          NaN  00:00-23:30;                  Sun;

[8078 rows x 29 columns]

In [17]:
# Cek missing value dengan fungsi isnull()
resto.isnull().sum()

userID               0
placeID              0
rating               0
food_rating          0
service_rating       0
Rpayment          5781
latitude          6917
longitude         6917
the_geom_meter    6917
name              6917
address           6917
city              6917
state             6917
country           6917
fax               6917
zip               6917
alcohol           6917
smoking_area      6917
dress_code        6917
accessibility     6917
price             6917
url               6917
Rambience         6917
franchise         6917
area              6917
other_services    6917
parking_lot       6917
hours             4619
days              4619
dtype: int64

In [18]:
# Menghitung jumlah rating, food_rating, dan service kemudian menggabungkannya berdasarkan placeID
resto.groupby('placeID').sum()

rating  food_rating  service_rating    latitude    longitude
placeID                                                              
132560       12           24               6   95.009216  -396.667653
132561       15           20              20   94.907276  -396.506024
132564       25           25              30   94.923698  -396.580739
132572       90           90              84  332.124707 -1514.890677
132583       24           24              30   75.689162  -396.937328
...         ...          ...             ...         ...          ...
135088       36           42              36  113.256068  -595.319338
135104       48           80              48  166.270875  -694.179039
135106       96           96              96  221.497088 -1009.760928
135108       65           65              55  243.498787 -1110.269437
135109       24           30              18   75.687139  -396.941400

[130 rows x 5 columns]

**Menggabungkan Data dengan Fitur Nama Resto**

In [19]:
# Definisikan dataframe rating ke dalam variabel all_resto_rate
all_resto_rate = rating
all_resto_rate

userID  placeID  rating  food_rating  service_rating
0     U1077   135085       2            2               2
1     U1077   135038       2            2               1
2     U1077   132825       2            2               2
3     U1077   135060       1            2               2
4     U1068   135104       1            1               2
...     ...      ...     ...          ...             ...
1156  U1043   132630       1            1               1
1157  U1011   132715       1            1               0
1158  U1068   132733       1            1               0
1159  U1068   132594       1            1               1
1160  U1068   132660       0            0               0

[1161 rows x 5 columns]

In [20]:
# Menggabungkan all resto_rate dengan dataframe geo berdasarkan placeID
all_resto_name = pd.merge(all_resto_rate, geo[['placeID','name']], on='placeID', how='left')
 
# Print dataframe all_resto_name
all_resto_name

userID  placeID  ...  service_rating                                   name
0     U1077   135085  ...               2                 Tortas Locas Hipocampo
1     U1077   135038  ...               1                  Restaurant la Chalita
2     U1077   132825  ...               2                        puesto de tacos
3     U1077   135060  ...               2                Restaurante Marisco Sam
4     U1068   135104  ...               2                                   vips
...     ...      ...  ...             ...                                    ...
1156  U1043   132630  ...               1                             palomo tec
1157  U1011   132715  ...               0                   tacos de la estacion
1158  U1068   132733  ...               0                          Little Cesarz
1159  U1068   132594  ...               1     tacos de barbacoa enfrente del Tec
1160  U1068   132660  ...               0  carnitas mata calle Emilio Portes Gil

[1161 rows x 6 columns]

**Menggabungkan Data dengan Fitur masakan Resto**

Menggabungkan variabel all_resto_name yang kita peroleh dari tahapan sebelumnya dengan fitur masakan resto (cuisine). Tujuannya, agar kita mengetahui masakan yang disediakan oleh restoran.

In [21]:
# Menggabungkan dataframe cuisine dengan all_resto_name dan memasukkannya ke dalam variabel all_resto 
all_resto = pd.merge(all_resto_name, cuisine, on='placeID', how='left')
all_resto

userID  placeID  ...                                   name   Rcuisine
0     U1077   135085  ...                 Tortas Locas Hipocampo  Fast_Food
1     U1077   135038  ...                  Restaurant la Chalita        NaN
2     U1077   132825  ...                        puesto de tacos    Mexican
3     U1077   135060  ...                Restaurante Marisco Sam    Seafood
4     U1068   135104  ...                                   vips    Mexican
...     ...      ...  ...                                    ...        ...
1326  U1043   132630  ...                             palomo tec    Mexican
1327  U1011   132715  ...                   tacos de la estacion    Mexican
1328  U1068   132733  ...                          Little Cesarz   Pizzeria
1329  U1068   132594  ...     tacos de barbacoa enfrente del Tec    Mexican
1330  U1068   132660  ...  carnitas mata calle Emilio Portes Gil        NaN

[1331 rows x 7 columns]

#Data Preparation

**Mengatasi Missing Value**

In [22]:
# Mengecek missing value pada dataframe all_resto
all_resto.isnull().sum()

userID              0
placeID             0
rating              0
food_rating         0
service_rating      0
name                0
Rcuisine          288
dtype: int64

In [23]:
# Membersihkan missing value dengan fungsi dropna()
all_resto_clean = all_resto.dropna()
all_resto_clean

userID  placeID  ...                                name   Rcuisine
0     U1077   135085  ...              Tortas Locas Hipocampo  Fast_Food
2     U1077   132825  ...                     puesto de tacos    Mexican
3     U1077   135060  ...             Restaurante Marisco Sam    Seafood
4     U1068   135104  ...                                vips    Mexican
5     U1068   132740  ...       Carreton de Flautas y Migadas    Mexican
...     ...      ...  ...                                 ...        ...
1325  U1043   132732  ...                  Taqueria EL amigo     Mexican
1326  U1043   132630  ...                          palomo tec    Mexican
1327  U1011   132715  ...                tacos de la estacion    Mexican
1328  U1068   132733  ...                       Little Cesarz   Pizzeria
1329  U1068   132594  ...  tacos de barbacoa enfrente del Tec    Mexican

[1043 rows x 7 columns]

In [24]:
# Mengecek kembali missing value pada variabel all_resto_clean
all_resto_clean.isnull().sum()

userID            0
placeID           0
rating            0
food_rating       0
service_rating    0
name              0
Rcuisine          0
dtype: int64

**Menyamakan Jenis masakan**

Cek ulang data setelah proses cleaning pada tahap sebelumnya

In [25]:
# Mengurutkan resto berdasarkan PlaceID kemudian memasukkannya ke dalam variabel fix_resto
fix_resto = all_resto_clean.sort_values('placeID', ascending=True)
fix_resto

userID  placeID  ...                        name   Rcuisine
1303  U1087   132560  ...          puesto de gorditas   Regional
1288  U1050   132560  ...          puesto de gorditas   Regional
14    U1067   132560  ...          puesto de gorditas   Regional
42    U1082   132560  ...          puesto de gorditas   Regional
1052  U1013   132572  ...                Cafe Chaires  Cafeteria
...     ...      ...  ...                         ...        ...
438   U1024   135106  ...  El Rincón de San Francisco    Mexican
178   U1020   135109  ...                   Paniroles    Italian
1071  U1041   135109  ...                   Paniroles    Italian
99    U1030   135109  ...                   Paniroles    Italian
180   U1051   135109  ...                   Paniroles    Italian

[1043 rows x 7 columns]

In [26]:
# Mengecek berapa jumlah fix_resto
len(fix_resto.placeID.unique())

95

In [27]:
# Mengecek kategori masakan yang unik
fix_resto.Rcuisine.unique()

array(['Regional', 'Cafeteria', 'American', 'Mexican', 'Fast_Food',
       'Italian', 'Armenian', 'Pizzeria', 'Japanese', 'Vietnamese',
       'Family', 'International', 'Game', 'Burgers', 'Bakery', 'Bar',
       'Breakfast-Brunch', 'Bar_Pub_Brewery', 'Mediterranean',
       'Cafe-Coffee_Shop', 'Contemporary', 'Seafood', 'Chinese'],
      dtype=object)

In [28]:
# Mengecek kategori masakan Game 
fix_resto[fix_resto['Rcuisine'] == 'Game']

userID  placeID  rating  food_rating  service_rating name Rcuisine
781   U1015   132851       1            1               1  KFC     Game
509   U1052   132851       1            0               2  KFC     Game
708   U1008   132851       1            1               1  KFC     Game
770   U1037   132851       2            2               1  KFC     Game
574   U1069   132851       1            0               0  KFC     Game
1188  U1131   132851       2            2               2  KFC     Game
764   U1111   132851       2            1               0  KFC     Game

In [29]:
# Mengecek kategori masakan pada nama restoran KFC
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC      Game
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC      Game
708   U1008   132851       1            1               1  KFC      Game
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC      Game
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC      Game
1188  U1131   132851       2            2               2  KFC      Game
764   U1111   132851       2            1               0  KFC      Game
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

KFC lebih cocok disebut sebagai restoran dengan kategori masakan American. Jadi, kita perlu mengganti kategori ‘Game’ dengan ‘American’. 

In [30]:
# Mengubah nama kategori masakan ‘Game’ menjadi ‘American’
fix_resto = fix_resto.replace('Game', 'American')
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC  American
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC  American
708   U1008   132851       1            1               1  KFC  American
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC  American
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC  American
1188  U1131   132851       2            2               2  KFC  American
764   U1111   132851       2            1               0  KFC  American
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

Dalam sistem rekomendasi yang kita kembangkan, penting untuk memastikan satu restoran mewakili satu kategori masakan.

In [31]:
# Membuat variabel preparation yang berisi dataframe fix_resto kemudian mengurutkan berdasarkan placeID
preparation = fix_resto
preparation.sort_values('placeID')

userID  placeID  ...                        name   Rcuisine
1303  U1087   132560  ...          puesto de gorditas   Regional
1288  U1050   132560  ...          puesto de gorditas   Regional
14    U1067   132560  ...          puesto de gorditas   Regional
42    U1082   132560  ...          puesto de gorditas   Regional
184   U1055   132572  ...                Cafe Chaires  Cafeteria
...     ...      ...  ...                         ...        ...
1224  U1002   135106  ...  El Rincón de San Francisco    Mexican
99    U1030   135109  ...                   Paniroles    Italian
178   U1020   135109  ...                   Paniroles    Italian
1071  U1041   135109  ...                   Paniroles    Italian
180   U1051   135109  ...                   Paniroles    Italian

[1043 rows x 7 columns]

Hanya perlu menggunakan data unik untuk dimasukkan ke dalam proses pemodelan. Oleh karena itu, kita harus menghapus data yang duplikat dengan fungsi drop_duplicates().

In [32]:
# Membuang data duplikat pada variabel preparation
preparation = preparation.drop_duplicates('placeID')
preparation

userID  placeID  ...                                name   Rcuisine
1303  U1087   132560  ...                  puesto de gorditas   Regional
1052  U1013   132572  ...                        Cafe Chaires  Cafeteria
168   U1118   132583  ...                    McDonalds Centro   American
24    U1107   132584  ...                  Gorditas Dona Tota    Mexican
1329  U1068   132594  ...  tacos de barbacoa enfrente del Tec    Mexican
...     ...      ...  ...                                 ...        ...
681   U1095   135086  ...         Mcdonalds Parque Tangamanga  Fast_Food
175   U1020   135088  ...                   Cafeteria cenidet  Cafeteria
4     U1068   135104  ...                                vips    Mexican
488   U1004   135106  ...          El Rincón de San Francisco    Mexican
178   U1020   135109  ...                           Paniroles    Italian

[95 rows x 7 columns]

In [33]:
# Mengonversi data series ‘placeID’ menjadi dalam bentuk list
resto_id = preparation['placeID'].tolist()
 
# Mengonversi data series ‘Name’ menjadi dalam bentuk list
resto_name = preparation['name'].tolist()
 
# Mengonversi data series ‘Rcuisine’ menjadi dalam bentuk list
resto_cuisine = preparation['Rcuisine'].tolist()
 
print(len(resto_id))
print(len(resto_name))
print(len(resto_cuisine))

95
95
95


In [34]:
# Membuat dictionary untuk data ‘resto_id’, ‘resto_name’, dan ‘cuisine’
resto_new = pd.DataFrame({
    'id': resto_id,
    'resto_name': resto_name,
    'cuisine': resto_cuisine
})
resto_new

id                          resto_name    cuisine
0   132560                  puesto de gorditas   Regional
1   132572                        Cafe Chaires  Cafeteria
2   132583                    McDonalds Centro   American
3   132584                  Gorditas Dona Tota    Mexican
4   132594  tacos de barbacoa enfrente del Tec    Mexican
..     ...                                 ...        ...
90  135086         Mcdonalds Parque Tangamanga  Fast_Food
91  135088                   Cafeteria cenidet  Cafeteria
92  135104                                vips    Mexican
93  135106          El Rincón de San Francisco    Mexican
94  135109                           Paniroles    Italian

[95 rows x 3 columns]

#Model Development dengan Content Based Filtering

Cek kembali data yang kita miliki dan assign dataframe dari tahap sebelumnya ke dalam variabel data

In [35]:
data = resto_new
data.sample(5)

id          resto_name           cuisine
49  132955           emilianos   Bar_Pub_Brewery
21  132754     Cabana Huasteca           Mexican
23  132766              Mikasa          Japanese
3   132584  Gorditas Dona Tota           Mexican
48  132954        la parroquia  Breakfast-Brunch

TF-IDF Vectorizer digunakan pada sistem rekomendasi untuk menemukan representasi fitur penting dari setiap kategori masakan

In [36]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada data cuisine
tf.fit(data['cuisine']) 
 
# Mapping array dari fitur index integer ke fitur nama
tf.get_feature_names()

['american',
 'armenian',
 'bar',
 'bar_pub_brewery',
 'breakfast',
 'brunch',
 'burgers',
 'cafe',
 'cafeteria',
 'chinese',
 'coffee_shop',
 'contemporary',
 'family',
 'fast_food',
 'international',
 'italian',
 'japanese',
 'mexican',
 'pizzeria',
 'regional',
 'seafood',
 'vietnamese']

In [37]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(data['cuisine']) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape 

(95, 22)

Nilai 95 merupakan ukuran data dan 22 merupakan matrik kategori masakan.

In [38]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0., 0., 0., ..., 1., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [1., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]])

Mengidentifikasi Korelasi antara Restoran dengan Kategori Masakannya

In [39]:
# Membuat dataframe untuk melihat tf-idf matrix
# Kolom diisi dengan jenis masakan
# Baris diisi dengan nama resto
 
pd.DataFrame(
    tfidf_matrix.todense(), 
    columns=tf.get_feature_names(),
    index=data.resto_name
).sample(22, axis=1).sample(10, axis=0)

regional  ...  armenian
resto_name                                                    ...          
Restaurante Bar Fu-hao                                   0.0  ...       0.0
la perica hamburguesa                                    0.0  ...       0.0
Restaurant and Bar and Clothesline Carlos N Cha...       0.0  ...       0.0
Cafeteria cenidet                                        0.0  ...       0.0
El Rincón de San Francisco                               0.0  ...       0.0
Gordas de morales                                        0.0  ...       0.0
Little Cesarz                                            0.0  ...       0.0
El Herradero Restaurante and Bar                         0.0  ...       0.0
La Posada del Virrey                                     0.0  ...       0.0
tacos de la estacion                                     0.0  ...       0.0

[10 rows x 22 columns]

Menghitung Derajat Kesamaan (similarity degree) antar Restoran dengan Teknik Cosine Similarity

In [40]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cosine_sim = cosine_similarity(tfidf_matrix) 
cosine_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

Lihat matriks kesamaan setiap resto dengan menampilkan nama restoran dalam 5 sampel kolom (axis = 1) dan 10 sampel baris (axis=0).

In [41]:
# Membuat dataframe dari variabel cosine_sim dengan baris dan kolom berupa nama resto
cosine_sim_df = pd.DataFrame(cosine_sim, index=data['resto_name'], columns=data['resto_name'])
print('Shape:', cosine_sim_df.shape)
 
# Melihat similarity matrix pada setiap resto
cosine_sim_df.sample(5, axis=1).sample(10, axis=0)

Shape: (95, 95)


resto_name                          Unicols Pizza  ...  Gorditas Dona Tota
resto_name                                         ...                    
Dominos Pizza                                 0.0  ...                 0.0
Restaurante Bar Fu-hao                        0.0  ...                 0.0
tacos de barbacoa enfrente del Tec            0.0  ...                 1.0
la parroquia                                  0.0  ...                 0.0
cafe punta del cielo                          0.0  ...                 0.0
Gorditas Doa Gloria                           0.0  ...                 1.0
Paniroles                                     1.0  ...                 0.0
Restaurante Bar El Gallinero                  0.0  ...                 0.0
Chaires                                       0.0  ...                 0.0
Restaurant Las Mananitas                      0.0  ...                 0.0

[10 rows x 5 columns]

matriks di atas berukuran 95 restoran x 95 restoran. Tapi kita tidak bisa menampilkan semuanya. Oleh karena itu, kita hanya memilih 10 restoran pada baris vertikal dan 5 restoran pada sumbu horizontal.

Dari matriks di atas dapat kita pahami sebagai: Luna Cafe teridentifikasi sama (similar) dengan Abondance Restaurant Bar.

#Mendapatkan Rekomendasi

Membuat fungsi resto_recommendations dengan beberapa parameter sebagai berikut:

- Nama_resto : Nama restoran (index kemiripan dataframe).
- Similarity_data : Dataframe mengenai similarity yang telah kita definisikan sebelumnya.
- Items : Nama dan fitur yang digunakan untuk mendefinisikan kemiripan, dalam hal ini adalah ‘resto_name’ dan ‘cuisine’.
- k : Banyak rekomendasi yang ingin diberikan.

Memberikan sejumlah rekomendasi restoran pada pengguna yang diatur dalam parameter k

In [46]:
def resto_recommendations(nama_resto, similarity_data=cosine_sim_df, items=data[['resto_name', 'cuisine']], k=5):
    """
    Rekomendasi Resto berdasarkan kemiripan dataframe
 
    Parameter:
    ---
    nama_resto : tipe data string (str)
                Nama Restoran (index kemiripan dataframe)
    similarity_data : tipe data pd.DataFrame (object)
                      Kesamaan dataframe, simetrik, dengan resto sebagai 
                      indeks dan kolom
    items : tipe data pd.DataFrame (object)
            Mengandung kedua nama dan fitur lainnya yang digunakan untuk mendefinisikan kemiripan
    k : tipe data integer (int)
        Banyaknya jumlah rekomendasi yang diberikan
    ---
 
 
    Pada index ini, kita mengambil k dengan nilai similarity terbesar 
    pada index matrix yang diberikan (i).
    """
 
 
    # Mengambil data dengan menggunakan argpartition untuk melakukan partisi secara tidak langsung sepanjang sumbu yang diberikan    
    # Dataframe diubah menjadi numpy
    # Range(start, stop, step)
    index = similarity_data.loc[:,nama_resto].to_numpy().argpartition(
        range(-1, -k, -1))
    
    # Mengambil data dengan similarity terbesar dari index yang ada
    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    
    # Drop nama_resto agar nama resto yang dicari tidak muncul dalam daftar rekomendasi
    closest = closest.drop(nama_resto, errors='ignore')
 
    return pd.DataFrame(closest).merge(items).head(k)

Dengan menggunakan argpartition, kita mengambil sejumlah nilai k tertinggi dari similarity data (dalam kasus ini: dataframe cosine_sim_df). Kemudian, kita mengambil data dari bobot (tingkat kesamaan) tertinggi ke terendah. Data ini dimasukkan ke dalam variabel closest. Berikutnya, kita perlu menghapus nama_resto yang yang dicari agar tidak muncul dalam daftar rekomendasi. Dalam kasus ini, nanti kita akan mencari resto yang mirip dengan KFC, sehingga kita perlu drop nama_resto KFC agar tidak muncul dalam daftar rekomendais yang diberikan nanti.  

#Cari resto yang mirip dengan KFC

In [47]:
data[data.resto_name.eq('KFC')]

id resto_name   cuisine
31  132851        KFC  American

In [48]:
# Mendapatkan rekomendasi restoran yang mirip dengan KFC
resto_recommendations('KFC')

resto_name        cuisine
0                VIPS       American
1  tacos los volcanes       American
2     Pizzeria Julios       American
3             Sirlone  International
4    McDonalds Centro       American